### PR TITLE
Multipart mime 3487 v27

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/3487

Describe changes:
- convert HTTP to use new rust mime parser
- convert SMTP to use new rust mime parser

Follows #9759 with extracting recursive files/boundaries to be the same as C

### CODE REVIEW

**Comment by suricata-qa** [comment]
ERROR:
 

ERROR: QA failed on SURI_TLPW1_files_sha256.
  

Pipeline 16444

**Comment by catenacyber** [comment]
This PR has just one more file than master
```
sha256_base\_master < baseline  vs  test >  sha256_test_sort differ

	smtp 2f4e7dba21a31bde1192ca03b489a9bd47281a28e206b3dcf245082a491e8e0a
49153 baseline/SURI_TLPW1/sha256_base_master sha256_base\_master lines
49154 /tmp/report_16444/_SURI_TLPW1/sha256_test_sort sha256_test_sort lines
```

**Comment by victorjulien** [comment]
@jasonish can you review the Rust MIME code in this PR?

File: rust/src/mime/mime.rs [hunk 1]
```diff
@@ -0,0 +1,581 @@
...
+            escaping = false;
+        }
+    }
+    // should fail
+    let (input, value) = take_until("\"")(input)?;
+    let (input, _) = char('"')(input)?;
+    return Ok((input, value));
+}
+
+fn mime_parse_value_until(input: &[u8]) -> IResult<&[u8], &[u8]> {
```
Comment (jasonish) [inline rust/src/mime/mime.rs:53]: nit: until what?

File: rust/src/mime/mime.rs [hunk 2]
```diff
@@ -0,0 +1,581 @@
...
+pub(crate) const RS_MIME_MAX_TOKEN_LEN: usize = 255;
+
+#[derive(Debug)]
+enum MimeParserState {
+    Start = 0,
+    Header = 1,
+    HeaderEnd = 2,
+    Chunk = 3,
+    BoundaryWaitingForEol = 4,
+}
```
Comment (jasonish) [inline rust/src/mime/mime.rs:162]: nit: do these need values for comparison to C or something? Idtiomatically we'd just use their default and leave off the values.

File: rust/src/mime/mime.rs [hunk 3]
```diff
@@ -0,0 +1,581 @@
...
+    return Ok((input, (MimeParserState::Header, fileopen, errored)));
+}
+
+type NomTakeError<'a> = Err<nom7::error::Error<&'a [u8]>>;
+
+fn mime_consume_chunk<'a>(boundary: &[u8], input: &'a [u8]) -> IResult<&'a [u8], bool> {
+    let r: Result<(&[u8], &[u8]), NomTakeError> = take_until("\r\n")(input);
+    match r {
+        Ok((input, line)) => {
+            let (input2, _) = tag("\r\n")(input)?;
```
Comment (jasonish) [inline rust/src/mime/mime.rs:319]: Naming? I hit this naming issue with nom as well often.

File: rust/src/mime/mime.rs [hunk 4]
```diff
@@ -0,0 +1,581 @@
...
+        }
+    } else {
+        *buffer = std::ptr::null_mut();
+        *filename_len = 0;
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn SCMimeStateFree(ctx: &mut MimeStateHTTP) {
+    // Just unbox...
```
Comment (jasonish) [inline rust/src/mime/mime.rs:497]: I don't think this is new code, but these comments are wrong now. We're explicitly dropping, not just unboxing.

**Review by jasonish — APPROVED** [review]
The Rust looks good, just minor nits. I do like to see more unit tests when it comes to nom parsers rather than relying on coverage from other sources, but coverage seems pretty good.

File: rust/src/mime/mime.rs (hunk 1 above)
Comment (catenacyber) [reply rust/src/mime/mime.rs:53]: until semicolon ;-) renaming it

File: rust/src/mime/mime.rs (hunk 2 above)
Comment (catenacyber) [reply rust/src/mime/mime.rs:162]: not needed, removing values

File: rust/src/mime/mime.rs (hunk 3 above)
Comment (catenacyber) [reply rust/src/mime/mime.rs:319]: renaming to `next_line`. Should I add more comments ?

File: rust/src/mime/mime.rs (hunk 4 above)
Comment (catenacyber) [reply rust/src/mime/mime.rs:497]: ok

**Comment by catenacyber** [comment]
Replaced by https://github.com/OISF/suricata/pull/9841

**Pull request closed without merging**